### PR TITLE
feat: MutiText value type validation [DHIS2-14737]

### PR DESCRIPTION
**Pull request by abyot**
- MultiText value validation
- MultiText values are separated by a comma
- Apply strict option code validation `equalsIgonoreCase(code)` => `equals(code)`
- Tests

In addition error message for validation code E1125 is changed from 
`"Value {value} is not a valid option for {dataelement_uid or attribute_uid} {option_set_uid} in option set {comma_separated_option_codes}"`  
to
`"Value {value} is not a valid option code in option set {option_set_uid}"`

### CODE REVIEW

File: dhis-2/dhis-support/dhis-support-system/src/main/java/org/hisp/dhis/system/util/ValidationUtils.java [hunk 1]
```diff
@@ -520,21 +519,19 @@ public static String valueIsValid( String value, DataElement dataElement, boolea
 
         OptionSet options = dataElement.getOptionSet();
 
-        boolean isMultiText = valueType == MULTI_TEXT;
-
-        if ( isMultiText && options == null )
+        if ( valueType.isMultiText() && options == null )
```
Comment (teleivo) [inline dhis-2/dhis-support/dhis-support-system/src/main/java/org/hisp/dhis/system/util/ValidationUtils.java:522]: This code looks a lot like the code in dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/validation/validator/ValidationUtils.java not sure if we can/should make it reusable 🤷🏻

File: dhis-2/dhis-test-integration/src/test/java/org/hisp/dhis/tracker/validation/EventImportValidationTest.java [hunk 2]
```diff
@@ -118,6 +118,26 @@ void failValidationWhenTrackedEntityAttributeHasWrongOptionValue()
         assertHasOnlyErrors( importReport, ValidationCode.E1125 );
     }
 
+    @Test
+    void successWhenEventHasValidMultiTextValue()
```
Comment (teleivo) [inline dhis-2/dhis-test-integration/src/test/java/org/hisp/dhis/tracker/validation/EventImportValidationTest.java:122]: I think this duplicates tests unit tests in dhis-2/dhis-services/dhis-service-tracker/src/test/java/org/hisp/dhis/tracker/validation/validator/event/DataValuesValidatorTest.java

I would therefore remove this one in favor of the unit tests

File: dhis-2/dhis-test-integration/src/test/java/org/hisp/dhis/tracker/validation/TrackedEntityImportValidationTest.java [hunk 3]
```diff
@@ -92,6 +92,28 @@ void successValidationWhenTrackedEntityAttributeHasValidOptionValue()
         assertNoErrors( importReport );
     }
 
+    @Test
+    void failValidationWhenTrackedEntityAttributeHasWrongMultiTextValue()
```
Comment (teleivo) [inline dhis-2/dhis-test-integration/src/test/java/org/hisp/dhis/tracker/validation/TrackedEntityImportValidationTest.java:96]: I think this duplicates tests unit tests in dhis-2/dhis-services/dhis-service-tracker/src/test/java/org/hisp/dhis/tracker/validation/validator/trackedentity/AttributeValidatorTest.java

I would therefore remove this one in favor of the unit tests

**Review by teleivo — COMMENTED** [review]
* I think we should keep the unit tests you wrote for the validation changes and one e2e test for the success case. I would remove all other tests as they just duplicate the former ones. In case we break the validation we would now have 3-4 tests failing for the same reason if I am not mistaken 😄 

* Is there a common separator that we use for a particular type (data/metadata)? 🤔 Some things like attribute category options are separated by semicolons if I remember correctly. Some seem to be separated by commas.

File: dhis-2/dhis-e2e-test/src/test/java/org/hisp/dhis/tracker/importer/tei/TeiValidationTests.java [hunk 4]
```diff
@@ -211,6 +213,32 @@ public void shouldReturnErrorWhenAttributeWithOptionSetInvalid()
...
+            .array();
+
+        trackerActions.postAndGetJobReport( trackedEntities, new QueryParamsBuilder().add( "async=false" ) )
+            .validateErrorReport()
+            .body( "errorCode", hasItem( "E1125" ) )
+            .body( "trackerType", hasItem( "TRACKED_ENTITY" ) );
+    }
+
+    @Test
+    public void shouldReturnSuccessWhenAttributeMultiTextInvalid()
```
[the remark this answers is not in the document]
Comment (abyot) [reply dhis-2/dhis-e2e-test/src/test/java/org/hisp/dhis/tracker/importer/tei/TeiValidationTests.java:230]: Thanks, it should be `IsValid`

File: dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/validation/ValidationCode.java [hunk 5]
```diff
@@ -117,7 +117,7 @@
     E1122( "Missing required enrollment property: `{0}`." ),
     E1123( "Missing required event property: `{0}`." ),
     E1124( "Missing required relationship property: `{0}`." ),
-    E1125( "Value {0} is not a valid option for {1} {2} in option set {3}" ),
+    E1125( "Value `{0}` is not a valid option code in option set `{1}`" ),
```
[the remark this answers is not in the document]
Comment (abyot) [reply dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/validation/ValidationCode.java:120]: Yes, we have long list of codes ... for example ICD code.

File: dhis-2/dhis-support/dhis-support-system/src/main/java/org/hisp/dhis/system/util/ValidationUtils.java (hunk 1 above)
Comment (abyot) [reply dhis-2/dhis-support/dhis-support-system/src/main/java/org/hisp/dhis/system/util/ValidationUtils.java:522]: We have a lot where aggregate and tracker are not collaborating. We can take this in a TECH issue.

File: dhis-2/dhis-services/dhis-service-tracker/src/test/java/org/hisp/dhis/tracker/validation/validator/event/DataValuesValidatorTest.java [hunk 6]
```diff
@@ -782,6 +782,73 @@ void failValidationDataElementOptionValueIsInValid()
         assertHasError( reporter, event, ValidationCode.E1125 );
     }
 
+    @Test
+    void successValidationDataElementMultiTextOptionValueIsValid()
+    {
+
```
[the remark this answers is not in the document]
Comment (abyot) [reply dhis-2/dhis-services/dhis-service-tracker/src/test/java/org/hisp/dhis/tracker/validation/validator/event/DataValuesValidatorTest.java:788]: yes, done.

File: dhis-2/dhis-e2e-test/src/test/java/org/hisp/dhis/tracker/importer/tei/TeiValidationTests.java [hunk 7]
```diff
@@ -211,6 +213,32 @@ public void shouldReturnErrorWhenAttributeWithOptionSetInvalid()
             .body( "trackerType", hasItem( "TRACKED_ENTITY" ) );
     }
 
+    @Test
+    public void shouldReturnErrorWhenAttributeMultiTextInvalid()
```
[the remark this answers is not in the document]
Comment (abyot) [reply dhis-2/dhis-e2e-test/src/test/java/org/hisp/dhis/tracker/importer/tei/TeiValidationTests.java:217]: I think we should also keep the failing e2e test to show we are returning proper error codes that users can somehow understand.

[the remark this answers is not in the document]
Comment (abyot) [reply dhis-2/dhis-e2e-test/src/test/java/org/hisp/dhis/tracker/importer/tei/TeiValidationTests.java:217]: ok, then we need rewrite all our e2e tests where we remove tests for failing params.

[the remark this answers is not in the document]
Comment (teleivo) [reply dhis-2/dhis-e2e-test/src/test/java/org/hisp/dhis/tracker/importer/tei/TeiValidationTests.java:217]: If we have e2e tests that are duplicating unit tests asserting that we return a specific error code than I would vote for removing the e2e tests 👍🏻

File: dhis-2/dhis-test-integration/src/test/java/org/hisp/dhis/tracker/validation/TrackedEntityImportValidationTest.java (hunk 3 above)
Comment (abyot) [reply dhis-2/dhis-test-integration/src/test/java/org/hisp/dhis/tracker/validation/TrackedEntityImportValidationTest.java]: ok, removed.

File: dhis-2/dhis-test-integration/src/test/java/org/hisp/dhis/tracker/validation/EventImportValidationTest.java (hunk 2 above)
Comment (abyot) [reply dhis-2/dhis-test-integration/src/test/java/org/hisp/dhis/tracker/validation/EventImportValidationTest.java]: ok, removed.

File: dhis-2/dhis-e2e-test/src/test/java/org/hisp/dhis/tracker/importer/tei/TeiValidationTests.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (abyot) [reply dhis-2/dhis-e2e-test/src/test/java/org/hisp/dhis/tracker/importer/tei/TeiValidationTests.java]: ok, removed.

**Comment by abyot** [comment]
> * I think we should keep the unit tests you wrote for the validation changes and one e2e test for the success case. I would remove all other tests as they just duplicate the former ones. In case we break the validation we would now have 3-4 tests failing for the same reason if I am not mistaken smile
> * Is there a common separator that we use for a particular type (data/metadata)? thinking Some things like attribute category options are separated by semicolons if I remember correctly. Some seem to be separated by commas.

not sure. I checked comma is used for the multitext and I just followed that.

File: dhis-2/dhis-services/dhis-service-tracker/src/test/java/org/hisp/dhis/tracker/imports/validation/validator/event/DataValuesValidatorTest.java [hunk 8]
```diff
@@ -783,9 +781,73 @@ void failValidationDataElementOptionValueIsInValid()
...
     {
+        DataValue validDataValue = dataValue( "CODE,CODE1" );
+        DataValue nullDataValue = dataValue( null );
+
+        OptionSet optionSet = new OptionSet();
+        Option option = new Option();
+        option.setCode( "CODE" );
+        Option option1 = new Option();
+        option1.setCode( "CODE1" );
+        optionSet.setOptions( Arrays.asList( option, option1 ) );
```
Comment (teleivo) [inline dhis-2/dhis-services/dhis-service-tracker/src/test/java/org/hisp/dhis/tracker/imports/validation/validator/event/DataValuesValidatorTest.java:794]: nitpick: we should prefer `List.of` when possible see https://github.com/dhis2/wow-backend/blob/master/guides/java11_pitfalls.md#listof-as-a-replacement-for-arraysaslist-and-alike

**Review by teleivo — APPROVED** [review]

File: dhis-2/dhis-services/dhis-service-tracker/src/test/java/org/hisp/dhis/tracker/imports/validation/validator/event/DataValuesValidatorTest.java (hunk 8 above)
Comment (abyot) [reply dhis-2/dhis-services/dhis-service-tracker/src/test/java/org/hisp/dhis/tracker/imports/validation/validator/event/DataValuesValidatorTest.java]: ok, good. replaced.

**Review by enricocolasante — APPROVED** [review]

**Pull request merged**